OpenRA.Mods.Common/DiscordService.cs

```diff
@@ -0,0 +1,200 @@
+			{
+				SkipIdenticalPresence = false
+			};
+
+			client.OnJoin += OnJoin;
+			client.OnJoinRequested += OnJoinRequested;
+
+			// HACK: We need to set HasRegisteredUriScheme to bypassing the check that is done when calling SetPresence with a joinSecret.
+			// DiscordRpc lib expect us to register uri handlers with RegisterUriScheme(), we are doing it ourselves in our installers/launchers.
+			client.GetType().GetProperty("HasRegisteredUriScheme").SetValue(client, true);
+
+			client.SetSubscription(EventType.Join | EventType.JoinRequest);
+			client.Initialize();
+		}
+
+		void OnJoinRequested(object sender, JoinRequestMessage args)
+		{
+			var client = (DiscordRpcClient)sender;
+			client.Respond(args, true);
+		}
+
+		void OnJoin(object sender, JoinMessage args)
+		{
+			if (currentState != DiscordState.InMenu)
+				return;
+
+			var server = args.Secret.Split('|');
+			Game.RunAfterTick(() =>
+			{
+				Game.RemoteDirectConnect(new ConnectionTarget(server[0], int.Parse(server[1])));
+			});
+		}
+
+		void SetStatus(DiscordState state, string details = null, string secret = null)
+		{
+			if (currentState == state)
+				return;
+
+			if (Service.Value == null)
+				return;
+
+			string stateText;
+			DateTime? timestamp = null;
+			Party party = null;
+			Secrets secrets = null;
+
+			switch (state)
+			{
+				case DiscordState.InMenu:
+					stateText = "In menu";
+					break;
+				case DiscordState.InMapEditor:
+					stateText = "In Map Editor";
+					break;
+				case DiscordState.InSkirmishLobby:
+					stateText = "In Skirmish Lobby";
+					break;
+				case DiscordState.InMultiplayerLobby:
+					stateText = "In Multiplayer Lobby";
+					timestamp = DateTime.UtcNow;
+					party = new Party
+					{
+						ID = Secrets.CreateFriendlySecret(new Random()),
+						Size = 1,
+						Max = 1
+					};
+					secrets = new Secrets
+					{
+						JoinSecret = secret
+					};
+					break;
+				case DiscordState.PlayingMultiplayer:
+					stateText = "Playing Multiplayer";
+					timestamp = DateTime.UtcNow;
+					break;
+				case DiscordState.PlayingCampaign:
+					stateText = "Playing Campaign";
+					timestamp = DateTime.UtcNow;
+					break;
+				case DiscordState.WatchingReplay:
+					stateText = "Watching Replay";
+					timestamp = DateTime.UtcNow;
+					break;
+				case DiscordState.PlayingSkirmish:
+					stateText = "Playing Skirmish";
+					timestamp = DateTime.UtcNow;
+					break;
+				default:
+					throw new ArgumentOutOfRangeException("state", state, null);
+			}
+
+			var richPresence = new RichPresence
+			{
+				Details = details,
+				State = stateText,
+				Assets = new Assets
+				{
+					LargeImageKey = "large",
+					LargeImageText = Game.ModData.Manifest.Metadata.Title,
+				},
+				Timestamps = timestamp.HasValue ? new Timestamps(timestamp.Value) : null,
+				Party = party,
+				Secrets = secrets
+			};
+
+			client.SetPresence(richPresence);
+			currentState = state;
+		}
+
+		public void Dispose()
+		{
+			client.Dispose();
```

Comment:
As suspected this crashes on dedicated servers, since we never created a client:
```
Unbehandelte Ausnahme: System.NullReferenceException: Der Objektverweis wurde nicht auf eine Objektinstanz festgelegt.
   at OpenRA.Mods.Common.DiscordService.Dispose() in \OpenRA\OpenRA.Mods.Common\DiscordService.cs:Line 166.
   at OpenRA.Manifest.Dispose() in \OpenRA\OpenRA.Game\Manifest.cs:Line 250.
   at OpenRA.Server.Program.Main(String[] args) in \OpenRA\OpenRA.Server\Program.cs:Line 85.
```